### PR TITLE
Sidecar containers in neutron agents

### PR DESCRIPTION
**Pull request by slawqo**
Add support for sidecar containers in neutron_dhcp and neutron_ovn roles

### CODE REVIEW

File: roles/edpm_neutron_dhcp/molecule/default/verify.yml
Comment (jpodivin) [inline roles/edpm_neutron_dhcp/molecule/default/verify.yml:67]: I would have gone with fail, but this is also valid.

File: roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2
Comment (jpodivin) [inline roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2:29]: ORPHANS?

File: roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2
Comment (jpodivin) [inline roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2:33]: These will pass no matter what, even if the runtime fails for some reason. It may be a good idea to at least report that the container wasn't removed.

**Review by jpodivin — CHANGES_REQUESTED** [review]

File: roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2
Comment (slawqo) [reply roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2:33]: I don't want to change this logic now. It's like it was in Tripleo until now. I added TODO with Your comment about logging and I will want to get back to it later in the follow up PR. Would that be ok for You?

Comment (jpodivin) [reply roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2:35]: Leaving it for later is fine. As long as we have a note about the deficiency.

File: roles/edpm_neutron_metadata/molecule/default/converge.yml
Comment (karelyatin) [inline roles/edpm_neutron_metadata/molecule/default/converge.yml:25]: what's triggering this change now? which user was used before?

**Review by karelyatin — CHANGES_REQUESTED** [review]
some comments inline

File: roles/edpm_neutron_metadata/molecule/default/converge.yml
Comment (slawqo) [reply roles/edpm_neutron_metadata/molecule/default/converge.yml:25]: I think that previously we didn't use this variable in this role and that's why it wasn't needed.

File: roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2
Comment (booxter) [inline roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2:23]: I think --net= may not be necessary here since we are already inside the netns (as can be seen from line 20 and inspection of the logic here).

File: roles/edpm_neutron_dhcp/templates/wrappers/haproxy.j2
Comment (booxter) [inline roles/edpm_neutron_dhcp/templates/wrappers/haproxy.j2:22]: I don't know details but I suspect that the wrapper is always present in our images and so we could simplify this command a bit. This would also allow to remove `bash` from the final CLI below.

Also, isn't it brittle to detect a "mode of operation" by the contents of the (common?) container image? I don't think we run haproxy under systemd, no? What if our image gets the wrapper installed in it? Will we change the behavior?

(AFAIU the wrapper is removed in RPMs. From changelog in Fedora:

```
* Fri Jan 05 2018 Ryan O'Hara <rohara@redhat.com> - 1.8.3-2
 - Remove haproxy-systemd-wrapper
```

But still, this seems brittle.

File: roles/edpm_neutron_dhcp/templates/wrappers/haproxy.j2
Comment (booxter) [inline roles/edpm_neutron_dhcp/templates/wrappers/haproxy.j2:14]: The logic of wrappers is largely repetitive. I wonder if we could consolidate it somehow.

File: roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2
Comment (booxter) [inline roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2:47]: I wonder... why does dnsmasq need access to namespaces volume here? Same for /etc/neutron. Same for the other two mounts.

File: roles/edpm_neutron_dhcp/molecule/default/verify.yml
Comment (booxter) [inline roles/edpm_neutron_dhcp/molecule/default/verify.yml:71]: I believe names of tasks must start from upper-case (after lint-production enforcement lands)

File: .github/workflows/molecule.yaml
Comment (booxter) [inline .github/workflows/molecule.yaml:30]: Thanks for enabling the molecule.

As discussed elsewhere, I think we should add a check in molecule scenario that would validate sidecar are not cleaned up when agent(s) are down.

For this, you would 

- make sure the agent container is up (already done in test)
- create qdhcp- (or metadata equivalent) namespace from inside the agent container
- start dnsmasq (or metadata equivalent) via wrapper (calling /usr/local/bin/<service> from the agent container)
- stop the agent container
- check that qdhcp- (or metadata equivalent) namespace is up and the service process is up

File: roles/edpm_neutron_metadata/templates/wrappers/haproxy.j2
Comment (booxter) [inline roles/edpm_neutron_metadata/templates/wrappers/haproxy.j2:1]: Looks like this file is not just similar but identical to what is used in the other dhcp role. I wonder if we could use a symlink here.

**Review by booxter — CHANGES_REQUESTED** [review]
My main concern here is that this is not actually tested (I mean - the fact that sidecars are NOT cleaned up on agent down). I think molecule scenario should be expanded to validate it.

I think core folks are also forceful about new arg specs having their descriptions filled in.

The rest of my comments are more inspirational, but I would appreciate if you could address some of them, or at least capture some of the questions as TODOs or elsewhere e.g. in Jira.

**Comment by booxter** [comment]
And also to capture the context of our 1:1 discussion that is relevant to my comments here for other reviewers, there is a concern that the existing tripleo integration for sidecar survival may not do what it's supposed to do, at least if I interpret the relevant tobiko test case correctly:

https://opendev.org/x/tobiko/src/commit/5d04e37ab05c214fbfa36d1d299b2cb8d0c0025c/tobiko/tests/faults/neutron/test_agents.py#L792

**Comment by booxter** [comment]
/hold

As per Slawek, this is WIP.

**Review by jpodivin — COMMENTED** [review]
Good work. But I'm going to wait until others questions are answered before giving it ++

File: roles/edpm_neutron_metadata/templates/wrappers/haproxy.j2
Comment (slawqo) [reply roles/edpm_neutron_metadata/templates/wrappers/haproxy.j2:1]: It's now done in more "unified" way. I hope it's fine for You

File: roles/edpm_neutron_dhcp/tasks/install.yml
Comment (jpodivin) [inline roles/edpm_neutron_dhcp/tasks/install.yml:22]: Please revert this. The `ansible_user` var isn't safe, we need something that works all the time. You can change the user it's executed under instead.

**Review by jpodivin — CHANGES_REQUESTED** [review]

File: roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2
Comment (cmsj) [reply roles/edpm_neutron_dhcp/templates/wrappers/dnsmasq.j2:3]: @booxter The guidance we've received is that if an individual file we're copying (or copying non-trivial amounts of content from) contains a copyright/attribution notice, we should retain that in our new file. Additionally, if the source repository's README contains a global notice of copyright/attribution, we should retain that in our README.

If, on the other hand, neither of those things are true (which I believe is the case with at least this particular file), then we do not need to reconstruct a copyright/attribution listing from the source repo's history (which is good because there's like 160 committers to puppet-tripleo!)

File: roles/edpm_neutron_dhcp/tasks/install.yml
Comment (slawqo) [reply roles/edpm_neutron_dhcp/tasks/install.yml:22]: Done

File: roles/edpm_neutron_metadata/molecule/default/verify.yml
Comment (booxter) [inline roles/edpm_neutron_metadata/molecule/default/verify.yml:19]: haven't you switched to testinfra? Should you not kill this script as you did for the other role?

File: roles/edpm_neutron_metadata/templates/wrappers/common_part.j2
Comment (booxter) [inline roles/edpm_neutron_metadata/templates/wrappers/common_part.j2:6]: shouldn't the name be different depending on the actual process?

File: roles/edpm_neutron_metadata/templates/wrappers/haproxy.j2
Comment (booxter) [inline roles/edpm_neutron_metadata/templates/wrappers/haproxy.j2:17]: the comment is not valid for reasons I listed elsewhere (no CLI args are checked to extract namespace name), but here also because you list dnsmasq in haproxy wrapper.

**Comment by booxter** [comment]
/lgtm

This is great!

**Pull request merged**
